test/test_utils/security.py

```diff
@@ -0,0 +1,189 @@
+
+
+class ECRScanFailureException(Exception):
+    """
+    Base class for other exceptions
+    """
+
+    pass
+
+
+class CVESeverity(IntEnum):
+    UNDEFINED = 0
+    INFORMATIONAL = 1
+    LOW = 2
+    MEDIUM = 3
+    HIGH = 4
+    CRITICAL = 5
+
+
+class ScanVulnerabilityList:
+    """
+    ScanAllowList is a class that reads an OS vulnerability allow-list, in the format stored on the DLC repo,
+    to allow easy comparison of any ECR Scan Vulnerabilities on an image with its corresponding allow-list file.
+    """
+
+    def __init__(self, minimum_severity=CVESeverity["MEDIUM"]):
+        self.vulnerability_list = {}
+        self.minimum_severity = minimum_severity
+
+    def construct_allowlist_from_file(self, file_path):
+        """
+        Read JSON file and prepare the object with all allowed vulnerabilities
+
+        :param file_path: Path to the allow-list JSON file.
+        :return: dict self.vulnerability_list
+        """
+        with open(file_path, "r") as f:
+            file_allowlist = json.load(f)
+        for package_name, package_vulnerability_list in file_allowlist.items():
+            for vulnerability in package_vulnerability_list:
+                if CVESeverity[vulnerability["severity"]] >= self.minimum_severity:
+                    if package_name not in self.vulnerability_list:
+                        self.vulnerability_list[package_name] = []
+                    self.vulnerability_list[package_name].append(vulnerability)
+        return self.vulnerability_list
+
+    def construct_allowlist_from_ecr_scan_result(self, vulnerability_list):
+        """
+        Read a vulnerability list and construct the vulnerability_list
+
+        :param vulnerability_list: list ECR Scan Result results
+        :return: dict self.vulnerability_list
+        """
+        for vulnerability in vulnerability_list:
+            package_name = get_ecr_vulnerability_package_name(vulnerability)
+            if package_name not in self.vulnerability_list:
+                self.vulnerability_list[package_name] = []
+            if CVESeverity[vulnerability["severity"]] >= self.minimum_severity:
+                self.vulnerability_list[package_name].append(vulnerability)
+        return self.vulnerability_list
+
+    def __contains__(self, vulnerability):
+        """
+        Check if an input vulnerability exists on the allow-list
+
+        :param vulnerability: dict JSON object consisting of information about the vulnerability in the format
+                              presented by the ECR Scan Tool
+        :return: bool True if the vulnerability is allowed on the allow-list.
+        """
+        package_name = get_ecr_vulnerability_package_name(vulnerability)
+        if package_name not in self.vulnerability_list:
+            return False
+        for allowed_vulnerability in self.vulnerability_list[package_name]:
+            if are_vulnerabilities_equivalent(vulnerability, allowed_vulnerability):
```

Comment:
is this is checking if vulnerability is allowed and can be skipped? If yes are_vulnerabilities_equivalent seems misleading, can be more like are_vulnerabilities_allowed?

Reply:
This just invokes a helper function that checks for "equivalence" between the vulnerability on the allow list and a vulnerability that is given as input for comparison.

The `are_vulnerabilities_equivalent` function only checks a subset of keys within the two dict objects being compared, which is why the function's name is `are_vulnerabilities_equivalent`.

`are_vulnerabilities_allowed` is a more abstract representation of what this class does, and is something that the pytest test function would check by running `if vulnerability in ScanVulnerabilityList_object:`